transformer_engine/pytorch/graph.py

```diff
@@ -408,6 +415,11 @@ def functionalized(*user_args, **user_kwargs):
 
                 skip_fp8_weight_update = not user_kwargs["is_first_microbatch"]
 
+            # Handle custom DDP from mcore
+            fuse_wgrad_accumulation = None
+            if "fuse_wgrad_accumulation" in user_kwargs:
+                fuse_wgrad_accumulation = user_kwargs["fuse_wgrad_accumulation"]
```

Comment:
Hey Tim, thanks for your comment.

I think `param_grad_added_to_main_grad` may be only toggled at [L435 which you mentioned](https://github.com/NVIDIA/Megatron-LM/blob/162b82df895113f75b55d1abb31d9c1ddaee223a/megatron/core/distributed/distributed_data_parallel.py#L435) and when the then wgrad fusion occurs [like here](https://github.com/NVIDIA/TransformerEngine/blob/a44cb72439574973809449e596464174e5a6fe37/transformer_engine/pytorch/module/linear.py#L582)

If we delete the first then wouldn't `param_grad_added_to_main_grad` only be necessarily set once when we trace the graph? And we are relying on all future steps that `param_grad_added_to_main_grad` hasn't changed and doesnt need to change. To me that seems like a large change in behavior.

When a user cudagraphs a module, I think the expectation is for no behavior to change, just give a cudagraphed version of the module, so my opinion is to match is as close as possible what eager mode does, which is to toggle `param_grad_added_to_main_grad` each step.

Reply:
On second thought maybe its better to move this PR into mcore entirely; since `param_grad_added_to_main_grad` only exists from Mcore, so if a user is running into this issue, they're already using MCore's cudagraph feature.